beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+/// @tparam GridCellDataMapType Type of the grid cell data map.
+/// @tparam Range Type of the states range.
+/// @tparam Weights Type of the weights range.
+/// @tparam Hashes Type of the hashes range.
+/// @param states Range containing the states of the particles.
+/// @param weights Range containing the weights of the particles.
+/// @param hashes Range containing the hashes of the particles.
+/// @return New instance of the grid cell data map populated with the information from the states.
+template <class GridCellDataMapType, class Range, class Weights, class Hashes>
+[[nodiscard]] auto populate_grid_cell_data_from_particles(Range&& states, Weights&& weights, const Hashes& hashes) {
+  GridCellDataMapType grid_cell_data;
+
+  // preallocate memory with a very rough estimation of the number of grid_cells we might end up with
+  grid_cell_data.reserve(states.size() / 5);
+
+  // calculate the accumulated cell weight and save a single representative_pose_in_world for each cell
+  for (const auto& [state, weight, hash] : ranges::views::zip(states, weights, hashes)) {
+    auto [it, inserted] = grid_cell_data.try_emplace(hash, GridCellData{});
+    it->second.weight += weight;
+    it->second.num_particles++;
+    if (inserted) {
+      it->second.representative_pose_in_world = state;
+    }
+  }
+
+  // normalize the accumulated weight by the number of particles in each cell
+  // to avoid biasing the clustering algorithm towards cells that randomly end up
+  // with more particles than others.
+  for (auto& [hash, entry] : grid_cell_data) {
+    entry.weight /= static_cast<double>(entry.num_particles);  // num_particles is guaranteed to be > 0
+  }
+
+  return grid_cell_data;
+}
+
+/// @brief Calculate the weight threshold that corresponds to a given percentile of the weights.
+/// @tparam GridCellDataMapType Type of the grid cell data map.
+/// @param grid_cell_data The grid cell data map.
+/// @param threshold The percentile of the weights to calculate the threshold for (range: 0.0 to 1.0)
+/// @return Threshold value that corresponds to the given percentile of the weights.
+template <class GridCellDataMapType>
+[[nodiscard]] auto calculate_percentile_weight_threshold(GridCellDataMapType&& grid_cell_data, double threshold) {
+  const auto extract_weight_f = [](const auto& grid_cell) { return grid_cell.second.weight; };
+  auto weights = grid_cell_data | ranges::views::transform(extract_weight_f) | ranges::to<std::vector<double>>() |
+                 ranges::actions::sort;
+  return weights[static_cast<std::size_t>(static_cast<double>(weights.size()) * threshold)];
+}
+
+/// @brief Cap the weight of each cell in the grid cell data map to a given value.
+/// @tparam GridCellDataMapType Type of the grid cell data map.
+/// @param grid_cell_data The grid cell data map.
+/// @param weight_cap The maximum weight value to be assigned to each cell.
+template <class GridCellDataMapType>
```

Comment:
meta: Maybe we can have a `GridCellData` class with member methods and the underlying map container could be a template parameter, instead of having free function templates that are really coupled with the map value type.

Reply:
I won't be addressing this.